### ISSUE TITLE
Using R6 classes rather than S4 classes

### ISSUE THREAD

**Issue by AlexPars** [stan-dev/rstan#514]
This is a good project, and I think that using R6 classes rather than S4 classes would enhance it from perspective of computational speed and future development.

NOTE: This is a suggestion rather than an issue.

**Comment by jgabry** [comment]
Thanks for the suggestion. We are definitely planning to move to reference
classes when we do the next major upgrades to the interfaces:

https://github.com/stan-dev/stan/wiki/User-Interface-Guidelines-for-Developers

**Comment by AlexPars** [comment]
Based on what I have read (and also on my own programming experience), R6 is a much better implementation in R of object-oriented paradigm compared to RC (Reference classes) , where better is in terms of computational performance, ease of usage and debug, encapsulation and inheritance

For example,  https://cran.r-project.org/web/packages/R6/vignettes/Performance.html

**Comment by bgoodri** [comment]
The performance differences are basically moot because more than 99.9% of
the time is going to be spent on the C++ side doing the computations. The
main points in favor of ReferenceClasses are static typing and locking. I
think the main point against was inheriting in other packages, but I think
that has been mitigated. Anyway, it is not decided and actually not even
that close to happening, but there is a rstan3 folder in the rstan GitHub
repo where I have done some stuff.

**Comment by jgabry** [comment]
Yeah I’m a fan of R6. I’ve used it a few times in packages I never ended up
releasing, but I liked using R6. I’m not sure if @bgoodri has already
considered it or not, but we’ll definitely look at many options including
R6 before we finalize anything.

**Comment by aaronjg** [comment]
For what it's worth, I'd vote to moving to the simpler S3 classes rather than reference classes/R6. Reference classes aren't really that widely used in the R ecosystem and are pretty unfamiliar to the average R user. RStan isn't using any of the fancy features of S4, like multiple dispatch, so S3 should be sufficient. 

Message passing OO isn't really used in R, so when it comes up, it is rather confusing to R users. R's class system is more similar to Julia's. However, the JuliaStan package is just a wrapper for cmdstan, so I'm not sure that's a great point of comparison.

I understand that using Reference Classes/R6 lets the interfaces be more consistent between C++, PyStan and RStan which makes development easier, but as someone who interfaces with Stan primarily through RStan, it would be nicer if the RStan package was more "R-like" and would be happy to help move things in that direction if that's where we want to take it.

**Comment by bgoodri** [comment]
All of the R6 / ReferenceClasses is speculative but thinking towards Stan
3.x when their features will be more relevant. I don't think going toward
S3 at this point makes sense.

**Comment by AlexPars** [comment]
Indeed I would not advocate using S4 or RC (Reference Classes) over S3, since the advantage/disadvantage ratio for S4 vs S3 or RC vs S3 is significantly tilted towards S3.

Situation is different for R6, which is a much better implementation of object oriented paradigm. That is why R6 is heavily used in RStudio packages and applications, including Shiny. 

S3 classes are great for prototyping and for relatively straightforward calculations. Thus I agree with bgoodri that  going toward S3 for future development would not be very useful.

**Comment by aaronjg** [comment]
I agree that R6 would be a better choice than RC but I'm not sure RStan needs that complexity. Building a webserver has very different requirements in terms of need for pass by reference than something like RStan. I disagree that R6 is a "better" implementation of OO-paradigm than S4. R6 is a better class-oriented paradigm, which is familiar to Java and C++ programmers, but S4 is a perfectly fine class-oriented paradigm and R users are familiar with. (S3 is okay for many uses too, a lot simpler than S4, but doesn't support multiple dispatch. [There's a nice comparison here](https://adv-r.hadley.nz/oo-tradeoffs.html).)

It probably depends on how we want the interface to look. 
Case 1:
```
stan_object <- compile_stan_model("model.stan")
stan_object$add_data(data)
stan_object$set_sampler_parameters(iter=1000, adapt_delta = 0.8, seed = 10)
stan_object$simulate()
theta <- stan_object$draws$theta
```
Case 2:
```
compiled_model <- compile_stan_model("model.stan")
model_with_data <- add_data(compiled_model, data)
stan_results <- sampling(model_with_data, iter=1000, ...)
theta <- extract(stan_results, theta)
```
In either case there could be a all-in-one convenience method `stan(model_code,data,algorithm)` that wraps everything.

If we want to go with the first case, R6 makes sense. However, that syntax feels very awkward as an R user, since calling functions on objects is not a very common paradigm in R. 

The second syntax is much more familiar paradigm in R. (Similarly, it would feel weird to have the second syntax in PyStan). @AlexPars, correct me if I'm wrong, but I don't think there is much advantage to using R6 if Stan is invoked using the case 2 semantics? You suggested this in the context of current RStan, not RStan3, so maybe I'm missing something? 

The second syntax also makes it easier for use cases such as calling the same model on different data since you could just add `model_with_data2 <- add_data(compiled_model, data2)` and proceed rather than adding doing `stan_object2 <- stan_object$copy()` before adding the data, or calling `compile_stan_model` a second time searching the workspace for already compiled models.

So in advocating S4 over using reference classes, I am mostly arguing to keep with more traditional R method oriented paradigm, rather than class oriented paradigm, since this is more familiar to R users.  People who are more comfortable with the class oriented syntax coming from Python will probably be using PyStan anyway.
